components/constellation/constellation.rs

```diff
@@ -1042,6 +1049,51 @@ where
             FromCompositorMsg::Keyboard(key_event) => {
                 self.handle_key_msg(key_event);
             },
+            // Perform a navigation previously requested by script, if approved by the embedder.
+            // If there is already a pending page (self.pending_changes), it will not be overridden;
+            // However, if the id is not encompassed by another change, it will be.
+            FromCompositorMsg::AllowNavigationResponse(pipeline_id, allowed) => {
+                let pending = self.pending_approval_navigations.remove(&pipeline_id);
+                let (browsing_context_id, top_level_browsing_context_id) = self
+                    .pipelines
+                    .get(&pipeline_id)
+                    .map(|pipeline| {
+                        (
+                            pipeline.browsing_context_id,
+                            pipeline.top_level_browsing_context_id,
+                        )
+                    })
+                    .expect("Navigated pipeline doesn't have a top_level_browsing_context");
+                match pending {
+                    Some((load_data, replace)) => {
+                        if allowed {
+                            self.load_url(
+                                top_level_browsing_context_id,
+                                pipeline_id,
+                                load_data,
+                                replace,
+                            );
+                        } else {
+                            // If the navigation is refused, and this concerns an iframe,
+                            // we need to take it out of it's "delaying-load-events-mode".
+                            let parent_pipeline_id =
+                                match self.browsing_contexts.get(&browsing_context_id) {
+                                    Some(browsing_context) => browsing_context.parent_pipeline_id,
+                                    None => return,
+                                };
+                            if let Some(_parent_pipeline_id) = parent_pipeline_id {
+                                self.handle_subframe_loaded(pipeline_id);
+                            }
```

Comment:
If the comment is correct, as in "concerns an iframe", I'm not sure about the matching logic. What exactly are you testing here? If the pipeline is part of a browsing context that is a top level browsing context?

Reply:
Ah yes, you're right, I've updated this to check whether the pipeline belongs to a top-level BC or not...